dll/win32/shell32/droptargets/CFSDropTarget.cpp

```diff
@@ -193,18 +222,21 @@ HRESULT CFSDropTarget::_GetEffectFromMenu(IDataObject *pDataObject, POINTL pt, D
 
     HMENU hpopupmenu = GetSubMenu(hmenu, 0);
 
+    SHELL_LimitDropEffectToItemAttributes(pDataObject, &dwAvailableEffects);
     if ((dwAvailableEffects & DROPEFFECT_COPY) == 0)
         DeleteMenu(hpopupmenu, IDM_COPYHERE, MF_BYCOMMAND);
-    else if ((dwAvailableEffects & DROPEFFECT_MOVE) == 0)
+    if ((dwAvailableEffects & DROPEFFECT_MOVE) == 0)
         DeleteMenu(hpopupmenu, IDM_MOVEHERE, MF_BYCOMMAND);
-    else if ((dwAvailableEffects & DROPEFFECT_LINK) == 0)
+    if ((dwAvailableEffects & DROPEFFECT_LINK) == 0 && (dwAvailableEffects & (DROPEFFECT_COPY | DROPEFFECT_MOVE)))
         DeleteMenu(hpopupmenu, IDM_LINKHERE, MF_BYCOMMAND);
 
-    if ((*pdwEffect & DROPEFFECT_COPY))
+    GetDefaultCopyMoveEffect();
+    // TODO: Why is this checking *pdwEffect?
```

Comment:
Looking at `CFSDropTarget::Drop()` who calls `_GetEffectFromMenu()`, we have:
```c
 399     DWORD dwAvailableEffects = *pdwEffect;
 400 
 401     _QueryDrop(dwKeyState, pdwEffect);
 402 
....
 407         HRESULT hr = _GetEffectFromMenu(pDataObject, pt, pdwEffect, dwAvailableEffects);
```
To me it looks like this:
- `*pdwEffect` given in input to `CFSDropTarget::Drop()` specifies the available effects wanted;
- this is stored into `dwAvailableEffects`
- Then, `_QueryDrop()` is called to retrieve the _actual_ default effect to select (now, the meaning of `*pdwEffect` changes);
- `_GetEffectFromMenu()` is then called with, pdwEffect such that _in input_, `*pdwEffect` specifies the default effect to select, while `dwAvailableEffects` specifies all the effects that are possible; and _ in output_, `*pdwEffect` receives the actual selected effect.

So, to my limited understanding, it would appear that the purpose of your new unimplemented `GetDefaultCopyMoveEffect()` helper is what `_QueryDrop()` helper tries to do.

Reply:
So I think that this `// TODO: Why is this checking *pdwEffect?` is answered, no?